Define Legacy RSE role and redistribute maintenance role skills and tasks into Project/community manager RSE role (#222)
CaptainSifff/paper_teaching-learning-RSE: competencies.md

The constantly evolving software environment can hinder or prevent reproducibility.
In this changing environment, a significant amount of effort in (research) software development
needs to be spent on maintenance to ensure that software remains useful or even installable.
Maintenance RSEs are mainly concerned with change management for code and documentation,
and working with the community to safeguard usability and adaptability.
They have additional competence in the governance of software projects,

> and scalable decision-making processes.

Reviewer: Is this really a specialization? I have the impression that we are already covering these in different parts of the paper. All RSEs aim to make code (re)usable, maintainable, etc.

I think that the proposed "Legacy RSEs" conveys the same message as the previous "Maintenance RSEs".

Author: @MakisH I think it is a specialization with regard to the software lifecycle.
I agree that RSEs strive to make their software (re)usable and maintainable, but this happens during the initial lifecycle phases, when the software is originally created.

Picking up from this previous work later in the lifecycle is the actual maintenance work, which not only happens later in the lifecycle, but also differs in tasks, where implementation may not be done by the RSE themself anymore, but rather by an already existing community. However, given an implicit or explicit *governance* model (which I think is an important aspect here), the work is not implementation, but rather integration, release engineering, documentation, evaluation, community work, etc.

This differs from Legacy RSE in that the nature (e.g. implementation language) or even the age of the codebase doesn't matter! Legacy RSE in contrast is decidedly concerned with doing this work for older codebases that use - let's say "more mature" - technologies such as the ones mentioned.

One could argue about the term "legacy" (i.e., doesn't software become legacy as soon as it is written?), or separating the two sections, but I think it's important to differentiate between maintenance RSE work (which is different from "generic" implementation-focused RSE work, and may perhaps happen in other institutional structures than RSE groups) and RSE work on older codebases.

So in a way, this PR is about describing the role more precisely in its facettes. I can merge the twoi sections again to include both if you prefer that? What do you think, @MakisH, @mhagdorn & @CaptainSifff?

Author: This may be a third role, but I was trying to describe another role in fact. Perhaps two examples may help that are close to what real RSEs do:

1. An application software has been developed by RSEs within a typical fixed-term research project to serve its own needs, has been well documented, published, etc. Now different external research groups (re-) use that software for their own projects. To be able to do their work, they want to make changes to the codebase, and use the application on a new OS version. One of the RSEs that have originally implemented the software has been tasked and is continuously funded (for, say, .3 FTE) within their department to maintain the software going forward. They work with the external research groups to discuss change requirements, improves documentation so that the external contributors can implement their solutions, reviews incoming code in PRs, asserts the engineering quality of the contributions through static code analysis metrics, makes merges, releases new versions of the software.

2. At a university's computing centre, a small team of RSEs is tasked with taking on the _maintenance_ of web applications (not just setting them up and running them on infrastructure) developed in research projects across the uni that have previously been run on a plethora of departmental, public and private servers. Initially, they work with the original PIs to understand the applications, their requirements, etc. They then adopt the applications on their central infrastructure, adapt them to run on the infrastructure, and update them (including on a the code level) as infrastructure changes and patches for libraries are being released. Eventually, they may make small changes to the logic as requested by researchers from their institution, or accept changes to the codebase from the same.

The second example could be used to extend the _Research infrastructure RSE_ role, but that depends on the *dimension* along which we want to describe the roles. At the moment, that dimension isn't clear, but perhaps it doesn't have to be.

The first example could be used to extend/sharpen the _Project/community manager RSE_ role. Alternatively, the role could be split up into _Community manager RSE_ (focusing on the people work) and _Maintenance RSE_ (focusing on the procedural and technical work). Both roles may often combine in one actual person, but they describe different aspects of RSE work.

Happy to hear others' thoughts about this.

Reviewer 2: So, how about focussing on Legacy RSEs and sharpening the Project Manager RSE in favour of elaborating the maintenance aspects in other places of the paper? E.g. in 4.4 the "Responsibilities section", and the project tables, there the notion of taking over code would fit into? @jcohen02

Reviewer 3: I definitely agree that a focus on the "Legacy RSE" specialisation seems reasonable - it seems likely that there will be RSEs who have specialist knowledge and skills in legacy programming languages, infrastructure tooling and even things like scientific instruments that can sometimes rely on control from a system that may be fixed to an outdated operating system and/or software stack.

In this context, maybe the Maintenance RSE specialisation could be renamed to Legacy RSE? (I think it was actually named this way in the first place and then got changed to Maintenance RSE?)

For the maintenance side of things, I'm wondering if this is more of a "cross-cutting" skill or requirement? i.e. rather than a specialisation in its own right, it's something that all RSEs doing any work with software are likely to need to do at some point. While there may be some software engineering skills that are more applicable to maintenance rather than the initial development of a piece of software, I feel that ultimately, maintenance leverages a wide ranging skill set that could equally be classed as a regular software engineering skill set.

In this context, I think my personal view would be more that that maintenance material might possible be better represented somewhere else, perhaps under a set of general skills that are applicable across many areas of the RSE skill space. I'm not sure quite where in the content this fits at the moment.

As I say, this is my perspective but It would be good to hear others' thoughts on this - @sdruskat, @MakisH, @CaptainSifff?

Reviewer: @sdruskat I think I understand what you are trying to convey, and it is definitely important. We do need people that have the skills to lead the maintenance operations of large projects (in terms of contributors), and not every research software project is a large one. I think the focus here is on the scalability of maintenance operations, not so much on the general maintenance, which happens on every kind of project. I think this needs to be more prominent for this distinct role to make complete sense.

And yes, it is distinct from the "Legacy RSE".

Reading the conversation, I was still wondering if this is a specialization or "just" a senior role (a manager, as @CaptainSifff wrote) that eventually comes. Does the "Maintenance RSE" learn something special at the M.Sc. RSE that not every other RSE learns? Can we have a "Junior Maintenance RSE" and a "Senior Maintenance RSE"? 

I think the name is problematic, because it has two meanings with subtle differences:
- maintaining something old so that it still works, despite no further big changes
- coordinating all the people that want to change the code and all related work packages

Thus, I think it is worth considering a different name. What would the "industry role" for such a person be called?

If I understand correctly, as a "Maintenance RSE", I don't develop new features myself or really do any research. I just oversee where the project is going (working together with other people that give directions) and make sure that people can contribute efficiently, PRs get merged (in the correct order), people don't step too much on each other's toes, the release notes make sense, I write porting guides, etc. Correct?

Author: @MakisH, @jcohen02, @CaptainSifff, thanks for your comments. I think we've reached agreement that "Legacy" work is indeed somewhat separate from general "Maintenance" work, and that the latter is more of a general skill set, which at least partly overlaps with what senior roles do?

I'll rework the changes to better reflect that, and will in the course of this try to find a sensible distribution of the maintenance skills over either existing roles or seniority or someplace else ;).

Author: Done in 69ff932e68e64ff30caee7e439cc36687b79ac07.

e.g., setup of virtual machines on hypervisors or the planning and setup of compute server clusters for \ac{ML}.
As an interface between the researchers and the infrastructure, they take care of user management, access permissions, and configuration of required services.

> #### Legacy RSEs

Reviewer 3: Tagging @sdruskat in case you don't get notified since this is already merged...

> #### Legacy RSEs

Reviewer 4: That's a really cool story and, yes, I very much like your addation

Research software may be used and evolved over generations of research without change management or governance processes, while software ecosystems (e.g., programming languages, frameworks, operating systems) constantly evolve.
This may lead to the emergence of legacy code that is actively used.
To safeguard continued usability and adoption,
these RSEs have experience in working with legacy code,

> and are competent in the application of software stacks that are not part of the general curricula any more (e.g., \gls{COBOL} or \gls{Fortran}).

Reviewer 4: just a minor thing, I think 

... that are no longer part of the general curricula (e.g....

reads a little better.

Author: ```suggestion
and are competent in the application of software stacks that are no longer part of the general curricula (e.g., \gls{COBOL} or \gls{Fortran}).
```
